### DISCUSSION TITLE
AI_Thinker Board - Arduino Configuration and  Metadata Live Capture Query

### DISCUSSION THREAD

**Discussion by waszee** [pschatzmann/arduino-audio-tools#1710] [Q&A] [answered]
I wanted to thank you for your examples using the AI Thinker A1S Auudio Kit board.  Like you I wanted to try some audio development without having all the jumper wires from my usual ESP development boards so purchased from Amazon an AI Thinker audio Lyrat clone. It came with no instructions or information and I was at a lost how to use.  Your posts saved the day.
 
I spent much of yesterday and this morning getting your examples to work.  Thought you might benefit from some feedback.  
 
First:  I think you should add somewhere in your AI Thinker Audio Kit description how to set the board in the Arduino IDE.   I had a lot of trouble with out of memory issues while trying to compile your examples.  I wonder how you set up your board.  Mine is now set up as:
 
ESP Dev Module
Huge APP (3M No OTA/1M SPIFFS)
Has PSRAM
 
I found the default memory settings limited text memory to about 1.2M and much of your code took 1.3M so had crashes.  Changing to the huge app memory configuration worked for me but still not sure it is the best option.  I will admit I was totally confused by the ESP-IDF Lyrat board set up when I tried to use the ESP-ADF.  Your Arduino approach was much easier. I also thought my board had 8MB from purchase info, but checking only found 4MB.  I am not sure of the PSRAM amount.
 
Second issue:
 
I want to port one of my internet radio ESP32-S3 ->i2S->MAX98357 to the AI Thinker-AIS->I2S->8388 output.    The S3 version uses the audio.h extensions to strip the metadata from running streams and move to a simple info structure where I find the Station Name when I change the URL to a new station.. The song Title is obtained each time a new song starts in the stream.   I output that information to a small i2c display screen (1” 128x64).  I also use a simple IR sensor with a low cost ELEGGO IR clicker to select channels and to adjust the amplifier volumes.  Expect to use the buttons instead of IR on the AI Thinker and attach a serial monitor if needed.
 
When I tried your metadata 2 example I see the initial header in the serial monitor but then get unwanted streaming blocks and no audio is output.   I do get audio from your stream-url-icy-audiokit example and was able to add the URLs that I like and use the next/prev buttons to move  thru the stations.  I am getting loud stutters{ “ da da da ….”]  while changing stations and code is not timing out if a connection fails so have to reboot sometimes.  I hope you and others might have some suggestions.

**Comment by pschatzmann** [comment, marked as answer]
The sketch size really depends on the sketch and you need to select the partition scheme accordingly with what functionality you want to use. Huge App is usually a good default setting.  I think RainMaker 4 MB no OTA will give you all 4 MB for your program.

Please google the topic PSRAM and ESP32: though it has 8MB you can only directly address 4 MB. 

The example sends the output to I2SStream. The AudioKit needs the codec to be set up to work, so just replace this with an AudioBoardStream and you should start to hear some sound. You can use the MetaDataType parameter in the callback to select the values that are relevant to you.

**Comment by waszee** [comment]
Thank you and will try the adjustments.  I have often ported my Arduino codes to platformio on VS Code so wonder if you might also have a platformio.ini configuration file for the AudioKit board to share? 

Also have been thinking about moving the network name and password from embedding in the code to the sd card.  The same with my list of streaming urls.  My kids and grandchildren live all over the states and I have to frequently change the network connection for my devices when visiting or vacationing. This currently requires a code edit and recompile so wonder how you or others recommend I handle.

**Reply by pschatzmann** [reply]
This (platformio) has been documented in the Wiki.

I think there are some cool libraries to manage the login

**Reply by waszee** [reply]
Made progress on the metadata2  and audio output but need help.  You were correct the audio  just needed lines to use the "kit" and change "i2s" to "kit: in a few places.

#include "AudioLibs/AudioBoardStream.h"
AudioBoardStream kit(AudioKitEs8388V1);
//EncodedAudioStream out2dec(&i2s, new MP3DecoderHelix()); // Decoding stream
EncodedAudioStream out2dec(&kit, new MP3DecoderHelix()); // Decoding stream
... 
I changed the input stream to look at folk alley which is a radio stream I often listen  and I know does a good job with posting metadata for the songs played.  I also added  url.httpRequest().header().put("Icy-MetaData","1"); based on the notes in the metadata example to get the ICY3data.  

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup multi output
  out.add(out1);
  out.add(out2dec);

  // setup input
  url.httpRequest().header().put("Icy-MetaData","1");
  url.begin( "http://freshgrass.streamguys1.com/folkalley-128mp3-tunein","audio/mp3");

  // setup metadata
  out1.setCallback(printMetaData);
  out1.begin(url.httpRequest());
...

I am now seeing the metadata that I  want as ==> type in the display below copied from my  Serial Monitor  but also a whole bunch of lines that I am not sure how to eliminate or where coming from, especial the StreamCopy.h lines that go on and on.


...
[I] HttpHeader.h : 253 - Data available: 2872
[I] HttpRequest.h : 164 - no CONTENT_LENGTH found in reply
[I] URLStream.h : 90 - size: 0
[I] URLStream.h : 97 - ==> http status: 200
[I] MetaDataICY.h : 267 - icy-metaint: 16000
[I] MetaDataICY.h : 278 - void audio_tools::ICYUrlSetup::executeCallback(void (*)(audio_tools::MetaDataType, const char*, int))
==> Genre: folk
==> Description: folkalley-128mp3
==> Name: folkalley-128mp3
[I] MetaDataICY.h : 63 - mp3_blocksize: 16000
[I] AudioTypes.h : 128 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[I] AudioTypes.h : 128 -  sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[I] I2SConfigESP32V1.h : 75 - rx/tx mode: TX_MODE
[I] I2SConfigESP32V1.h : 76 - port_no: 0
[I] I2SConfigESP32V1.h : 77 - is_master: Master
[I] I2SConfigESP32V1.h : 78 - sample rate: 44100
[I] I2SConfigESP32V1.h : 79 - bits per sample: 16
[I] I2SConfigESP32V1.h : 80 - number of channels: 2
[I] I2SConfigESP32V1.h : 81 - signal_type: Digital
[I] I2SConfigESP32V1.h : 82 - buffer_count:6
[I] I2SConfigESP32V1.h : 83 - buffer_size:512
[I] I2SConfigESP32V1.h : 85 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32V1.h : 88 - use_apll: true
[I] I2SConfigESP32V1.h : 94 - pin_mck: 0
[I] I2SConfigESP32V1.h : 96 - pin_bck: 27
[I] I2SConfigESP32V1.h : 98 - pin_ws: 25
[I] I2SConfigESP32V1.h : 100 - pin_data: 26
[I] I2SESP32V1.h : 205 - dma_frame_num: 768
[I] I2SESP32V1.h : 228 - tx: 26, rx: -1
[I] AudioEncoded.h : 75 - virtual void audio_tools::EncodedAudioOutput::addNotifyAudioChange(audio_tools::AudioInfoSupport&)
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioTypes.h : 128 - MP3DecoderHelix sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[I] I2SCodecStream.h : 92 - virtual void audio_tools::I2SCodecStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 128 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[I] I2SStream.h : 93 - virtual void audio_tools::I2SStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 128 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] MetaDataICY.h : 111 - Data ended
[I] MetaDataICY.h : 122 - metaDataLen: 80
[I] MetaDataICY.h : 128 - Metadata found
[I] MetaDataICY.h : 218 - StreamTitle='When You Miss Someone - William Prince';StreamUrl='';
==> Title: When You Miss Someone - William Prince
[I] MetaDataICY.h : 143 - Metadata ended
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops

**Reply by pschatzmann** [reply]
Just use an IcyStream instead of a HttpStream and defining the http parameter.
Do I understnad you right that you complain that you were setting the log level to Info instread of Warning in your sketch ?

**Reply by waszee** [reply]
I do not see an option to set the use of LOGI in the metadata2 sketch on my Arduino IDE.  The code says I need to find a define COPY_LOG_OFF but do not know where value set?  Not sure what you mean by IcyStream instead of HttpStream? Neither term is seen in the sketch so likely in a sub level but my level of understanding is too vague to dig out. 

lines 156-159 Streamcopy.h

#ifndef COPY_LOG_OFF
                LOGI("StreamCopy::copy %s %u -> %u -> %u bytes - in %u hops",log_name, (unsigned int)bytes_to_read,(unsigned int) bytes_read, (unsigned int)result, (unsigned int)delayCount);
                #endif
                //TRACED();

**Reply by pschatzmann** [reply]
I guess you define the logger in the first or second line in your setup(). Just read the Readme of this project or the Introduction Wiki
Here is the documentation for the [ICYStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_i_c_y_stream.html) class

**Reply by waszee** [reply]
Oh the AutoLogger line.  Did not know what it was doing.  Will check out.

**Reply by waszee** [reply]
Changing to the Warning Level was a huge fix for me. Looks like need to move to Error level to get rid of warnings.  I am hearing a chirping chatter in my audio.  Not sure what causing.  Will moving to the ICYStream fix?

==> Name: folkalley-128mp3
==> Title: Santiago - Tre Burt
==> Title: Polly Put the Kettle On-Lonesome Jo - Evie Ladin Band
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
==> Title: Deadlines and Dollar Signs - Patty Larkin - Tango
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono
[W] I2SESP32V1.h : 187 - Using channel_format: I2SChannelSelect::Left for mono

**Reply by pschatzmann** [reply]
It's just telling you that your music is in mono and will therefore play on the left channel only

**Reply by waszee** [reply]
Trying to split stream while using the audio player. The player without the MetaDataOutput lines worked but this is not working so seek clarification on how to best get the metadata output copied and metadata stream using the player connection..  

FYI did get Platformio to run arduino esamples.  Had to create a partition.csv to overcome the default flash memory limitation but otherwise worked.  Now back trying to strip metadata Titles.  

...
ICYStream urlStream(wifi, password);

MetaDataOutput out1;
MultiOutput out;
StreamCopy copier(out, urlStream(wifi,password)); // copy url to decoder

AudioSourceURL source(urlStream, urls, "audio/mp3");
AudioBoardStream kit(AudioKitEs8388V1);
MP3DecoderHelix decoder;
AudioPlayer player(source, kit, decoder);

...
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);

  // setup navigation
  kit.addAction(kit.getKey(4), next);
  kit.addAction(kit.getKey(3), previous);

//tried to add split stream or a second copy for metadata
  out.add(out1);
  out1.httpRequest().header().put("Icy-MetaData","1");
  out1.setCallback(printMetaData);
  player.setVolume(0.7);
  player.begin();
}
...

**Reply by pschatzmann** [reply]
Did you have a look at the class [documentation of the AudioPlayer](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_player.html) ? 

It directly provides a metadata callback!

**Comment by waszee** [comment]
Wow.  The last comment to look harder at the AutoPlayer documentation solved this issue for me.  Set AutoLogger to Warning Level, and added a single line to setup player section as well inserted the printMetaData subroutine from the metadata2 example.  This code is also playing non-chirping audio output which I suspect was from the autologger info level dumping too much out.
: 
 player.setMetadataCallback(printMetaData);

Example output shows a  few real errors and warnings.  I moved thru two US urls and the moved to liink in Switzerland.  That link, after a long pause, repaired and the program ran on.  Below shows the  serial moinitor output. 

..
==> Genre: folk
==> Description: folkalley-128mp3
==> Name: folkalley-128mp3
==> Title: Murder of Crows - Greensky Bluegrass
==> Title: A Case of You - Joni Mitchell - Both Sides Now
[W] URLStream.h : 342 - Redirected to: http://18093.live.streamtheworld.com/SP_R2135669_SC
==> Description: The Beauty of Sound
==> Name: The Scenic Route
==> Title: Yanni - If I Could Tell You
[159607][E][WiFiClient.cpp:275] connect(): socket error on fd 48, errno: 104, "Connection reset by peer"
[E] HttpRequest.h : 223 - Connect failed
[E] MetaDataICY.h : 269 - icy-metaint not defined
[W] ICYStream.h : 73 - url does not provide metadata
==> Genre: Entertainment
==> Description: SwissJazz_MP3@128Kbps
==> Name: /SwissJazz_Room1
==> Title: Erroll Garner - When You're Smiling

Thank you Phil Schatzmann!
